Firestore/core/src/firebase/firestore/local/query_data.h

```diff
@@ -0,0 +1,123 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_QUERY_DATA_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_QUERY_DATA_H_
+
+#include <cstdint>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/** An enumeration for the different purposes we have for queries. */
+enum class QueryPurpose {
+  /** A regular, normal query. */
+  kListen,
+
+  /**
+   * The query was used to refill a query after an existence filter mismatch.
+   */
+  kExistenceFilterMismatch,
+
+  /** The query was used to resolve a limbo document. */
+  kLimboResolution,
+};
+
+/**
+ * An immutable set of metadata that the store will need to keep track of for
+ * each query.
+ */
+class QueryData {
+ public:
+  /**
+   * Creates a new QueryData with the given values.
+   *
+   * @param query The query being listened to.
+   * @param target_id The target to which the query corresponds, assigned by the
+   *     LocalStore for user queries or the SyncEngine for limbo queries.
+   * @param purpose The purpose of the query.
+   * @param snapshot_version The latest snapshot version seen for this target.
+   * @param resume_token An opaque, server-assigned token that allows watching a
+   *     query to be resumed after disconnecting without retransmitting all the
+   *     data that matches the query. The resume token essentially identifies a
+   *     point in time from which the server should resume sending results.
+   */
+  QueryData(const core::Query& query,
+            int target_id,
+            QueryPurpose purpose,
+            const model::SnapshotVersion& snapshot_version,
+            const std::vector<uint8_t>& resume_token);
+
+  /**
+   * Convenience constructor for use when creating a QueryData for the first
+   * time.
+   */
+  // TODO(rsgowman): Define once WatchStream::EmptyResumeToken exists.
+  //QueryData(const core::Query& query, int target_id, QueryPurpose purpose);
+
+  const core::Query& query() const {
+    return query_;
+  }
+
+  int target_id() const {
+    return target_id_;
+  }
+
+  QueryPurpose purpose() const {
+    return purpose_;
+  }
+
+  const model::SnapshotVersion& snapshot_version() const {
+    return snapshot_version_;
+  }
+
+  const std::vector<uint8_t>& resume_token() const {
+    return resume_token_;
+  }
+
+  QueryData Copy(const model::SnapshotVersion& snapshot_version,
```

Comment:
Nitpick: the name doesn't fully represent what the function is doing, because it returns a modified copy. Perhaps something like `WithSnapshot` (`QueryData foo = bar.WithSnapshot(SnapshotVersion{"baz"}, token);`)? Feel free to ignore if this is for consistency with other platforms.

Reply:
Hm. The other platforms are inconsistent. Dumping this to the chat

...

Consensus seems to be to just use copy. I've left this as is.